pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,217 @@
 package certificate
 
 import (
-	"time"
-
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-
 	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
-	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/errcode"
 )
 
-// UseCase is how the certificate is used. Each use case will have its own rotation mechanism, that is necessary to
-// determine if we can propagate the rotation forward.
-type UseCase string
+func (m *Manager) handleMRCEvent(event MRCEvent) error {
+	log.Debug().Msgf("handling MRC event for MRC %s", event.MRCName)
+	mrcList, err := m.mrcClient.ListMeshRootCertificates()
+	if err != nil {
+		return err
+	}
 
-const (
-	// UseCaseValidatingWebhook maps to the certificate use case for the validating webhook.
-	UseCaseValidatingWebhook UseCase = "validatingWebhook"
+	filteredMRCList := filterOutInactiveMRCs(mrcList)
 
-	// UseCaseMutatingWebhook maps to the certificate use case for the Mutating Webhook
-	UseCaseMutatingWebhook UseCase = "mutatingWebhook"
+	desiredSigningMRC, desiredValidatingMRC, err := getSigningAndValidatingMRCs(filteredMRCList)
+	if err != nil {
+		return err
+	}
 
-	// UseCaseXDSControlPlane maps to the certificate use case for the XDS Control Plane
-	UseCaseXDSControlPlane UseCase = "xdsControlPlane"
+	m.mu.Lock()
+	signingIssuer := m.signingIssuer
+	validatingIssuer := m.validatingIssuer
+	m.mu.Unlock()
 
-	// UseCaseSidecar maps to the certificate use case for the  sidecar service certs
-	UseCaseSidecar UseCase = "sidecar"
+	// if the issuers are already set to the desired value, return
+	if signingIssuer != nil && signingIssuer.ID == desiredSigningMRC.Name && validatingIssuer != nil && validatingIssuer.ID == desiredValidatingMRC.Name {
+		return nil
+	}
 
-	// UseCaseBootstrap maps to the certificate use case for the  Bootstrap secrets
-	UseCaseBootstrap UseCase = "bootstrap"
+	desiredSigningIssuer, desiredValidatingIssuer, err := m.getCertIssuers(desiredSigningMRC, desiredValidatingMRC)
+	if err != nil {
+		return err
+	}
 
-	// UseCaseGateway maps to the certificate use case for the  Gateway ingress cert.
-	UseCaseGateway UseCase = "gateway"
-)
+	return m.updateIssuers(desiredSigningIssuer, desiredValidatingIssuer)
+}
+
+var validMRCIntentCombinations = map[v1alpha2.MeshRootCertificateIntent][]v1alpha2.MeshRootCertificateIntent{
+	v1alpha2.ActiveIntent: {
+		v1alpha2.PassiveIntent,
+		v1alpha2.ActiveIntent,
+	},
+	v1alpha2.PassiveIntent: {
+		v1alpha2.ActiveIntent,
+	},
+}
 
-var (
-	allUseCases = []UseCase{
-		UseCaseValidatingWebhook,
-		UseCaseMutatingWebhook,
-		UseCaseXDSControlPlane,
-		UseCaseSidecar,
-		UseCaseBootstrap,
-		UseCaseGateway,
+// ValidateMRCIntents validates the intent combination of MRCs
+func ValidateMRCIntents(mrc1, mrc2 *v1alpha2.MeshRootCertificate) error {
```

Comment:
super nit: by having the params be just the intent field, it makes it a bit more clear what this is evaluating without needing to read the code. More of an API hygiene thing. non-blocking.

Reply:
Updated!